core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -772,135 +793,159 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
+                  }
+                } else {
+                  // If the epoch is going to be fenced, it bumps the epoch differently with TV2.
+                  (!isFromClient || producerEpoch == txnMetadata.clientProducerEpoch) && producerEpoch >= txnMetadata.clientProducerEpoch
                 }
-              } else {
-                // If the epoch is going to be fenced, it bumps the epoch differently with TV2.
-                (!isFromClient || producerEpoch == txnMetadata.producerEpoch) && producerEpoch >= txnMetadata.producerEpoch
               }
-            }
-
-            val isRetry = retryOnEpochBump || retryOnOverflow
 
-            def generateTxnTransitMetadataForTxnCompletion(nextState: TransactionState, noPartitionAdded: Boolean): ApiResult[(Int, TxnTransitMetadata)] = {
-              // Maybe allocate new producer ID if we are bumping epoch and epoch is exhausted
-              val nextProducerIdOrErrors =
-                if (!isEpochFence && txnMetadata.isProducerEpochExhausted) {
-                  try {
-                    Right(producerIdManager.generateProducerId())
-                  } catch {
-                    case e: Exception => Left(Errors.forException(e))
+              val isRetry = retryOnEpochBump || retryOnOverflow
+
+              def generateTxnTransitMetadataForTxnCompletion(nextState: TransactionState, noPartitionAdded: Boolean): ApiResult[(Int, TxnTransitMetadata)] = {
+                // Maybe allocate new producer ID if the epoch is exhausted
+                // Note that we can arrive at this scenario when a client did a few keep-prepared calls and then
+                // decided to do one without keep-prepared (e.g. to force-terminate), so if we have
+                // nextProducerEpoch, then we need to do the epoch bump even for the isEpochFence case.
+                val nextProducerIdOrErrors =
+                  if ((!isEpochFence || txnMetadata.hasNextProducerEpoch) && txnMetadata.isProducerEpochExhausted) {
+                    try {
+                      Right(producerIdManager.generateProducerId())
+                    } catch {
+                      case e: Exception => Left(Errors.forException(e))
+                    }
+                  } else {
+                    Right(txnMetadata.nextProducerId)
                   }
+
+                // If the next producer epoch is set (which can happen if we used InitProducerId(keepPreparedTxn))
+                // then we also need to bump this epoch -- we want to fence requests that were issued with that
+                // epoch after the transaction is complete.  Consider this example:
+                //  1. (initial state): Ongoing, producerId = 42, producerEpoch = 100, nextProducerId = -1, nextProducerEpoch = -1
+                //  2. InitProducerId(keepPreparedTxn): Ongoing, producerId = 42, producerEpoch = 100, nextProducerId = 42, nextProducerEpoch = 101
```

Comment:
At step 2, are we returning the epoch 100 or 101 to the client? In other words, what is the epoch we send at step 3?